pkg/plugins/golang/v4/scaffolds/internal/templates/devcontainer.go

```diff
@@ -0,0 +1,153 @@
+/*
+Copyright 2022 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package templates
+
+import (
+	"sigs.k8s.io/kubebuilder/v4/pkg/machinery"
+)
+
+const devContainerTemplate = `// For format details, see https://aka.ms/devcontainer.json. For config options, see the
+// README at: https://github.com/devcontainers/templates/tree/main/src/docker-in-docker
+{
+  "name": "Kubebuilder DevContainer",
+  "build": {
+    "dockerfile": "./Dockerfile",
+    "context": "."
+  },
+
+  "features": {
+    "ghcr.io/devcontainers/features/docker-in-docker:2.11.0": {
+      "moby": true,
+      "azureDnsAutoDetection": true,
+      "version": "latest"
+    },
+    "ghcr.io/devcontainers/features/common-utils:2": {
+      "installZsh": false,
+      "upgradePackages": true,
+      "username": "devcontainer",
+      "userUid": "1001",
+      "userGid": "1001"
+    },
+    "ghcr.io/devcontainers/features/git:1": {
+      "ppa": true,
+      "version": "system"
+    },
+    "ghcr.io/devcontainers/features/go:1": {
+      "version": "1.22.0"
+    }
+  },
+
+  "runArgs": ["--network=host"],
+  // "forwardPorts": [],
+
+  "postCreateCommand": "bash ./.devcontainer/post-install.sh"
+
+  // Configure tool-specific properties.
+  // "customizations": {},
+
+  // Uncomment to connect as root instead. More info: https://aka.ms/dev-containers-non-root.
+  // "remoteUser": "root"
+}
+`
+
+const devContainerDockerfileTemplate = `# Build the manager binary
+# [Choice] Go version (use -bullseye variants on local arm64/Apple Silicon): 
+# 1, 1.19, 1.18, 1-bullseye, 1.19-bullseye, 1.18-bullseye, 1-buster, 1.19-buster, 1.18-buster
+ARG TAG=1.22
+FROM golang:${TAG}
+
+COPY post-install.sh .
+
+# Install dependencies
+RUN curl -LO "https://dl.k8s.io/release/$(curl -L -s https://dl.k8s.io/release/stable.txt)/bin/linux/amd64/kubectl" && \
+    install -o root -g root -m 0755 kubectl /usr/local/bin/kubectl
+
+RUN go install sigs.k8s.io/kind@v0.23.0 && \
+    curl -o kubebuilder \
+      -L https://github.com/kubernetes-sigs/kubebuilder/releases/latest/download/kubebuilder_linux_amd64 && \
+    chmod +x kubebuilder && \
+    mv kubebuilder /usr/local/bin/ && \
+    chmod +x post-install.sh
```

Comment:
We need add that in teh devContainer i.e.

```
 "postCreateCommand": "go mod tidy && make install && go version && kind version && kubebuilder version && docker --version",
  "onCreateCommand": [
    "curl -Lo ./kind https://kind.sigs.k8s.io/dl/latest/kind-linux-amd64",
    "chmod +x ./kind",
    "mv ./kind /usr/local/bin/kind",
    "curl -L https://github.com/kubernetes-sigs/kubebuilder/releases/latest/download/kubebuilder_linux_amd64 -o kubebuilder",
    "chmod +x kubebuilder",
    "mv kubebuilder /usr/local/bin/"
  ]
}
```

reason
- scaffold only 1 files instead of many
- have all logic and config to run the devContainer inside of its file to make easier the understand

Reply:
@camilamacedo86  Hey, I have time in moving the all of the commands from `post-install.sh`, and `Dockerfile` to devcontainer.json's `postCreateCommand` in array forma like this:
```json
"onCreateCommand": [
    // Update Aptitude repos
    "apt-get update",
    "apt-get upgrade"
]
```
 It fails every time due to this error:
```
OCI runtime exec failed: exec failed: unable to start container process: exec: "apt-get update": executable file not found in $PATH: unknown
[32470 ms] onCreateCommand failed with exit code 126. Skipping any further user-provided commands.
Error: Command failed: apt-get update apt-get upgrade
    at N7 (/home/linuxbrew/.linuxbrew/lib/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:235:130)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async Am (/home/linuxbrew/.linuxbrew/lib/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:227:4393)
    at async Xw (/home/linuxbrew/.linuxbrew/lib/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:227:3738)
    at async $w (/home/linuxbrew/.linuxbrew/lib/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:227:2745)
    at async fa (/home/linuxbrew/.linuxbrew/lib/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:227:2386)
    at async DtA (/home/linuxbrew/.linuxbrew/lib/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:465:1496)
    at async NH (/home/linuxbrew/.linuxbrew/lib/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:465:964)
    at async KtA (/home/linuxbrew/.linuxbrew/lib/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:482:3771)
    at async eB (/home/linuxbrew/.linuxbrew/lib/node_modules/@devcontainers/cli/dist/spec-node/devContainersSpecCLI.js:482:4886)
{"outcome":"error","message":"Command failed: apt-get update apt-get upgrade","description":"The onCreateCommand in the devcontainer.json failed.","containerId":"c4061bf17461ef8388a709c994a6280a249b14c6a30d3a6b203956ffe264bc6c"}
```

I've verified multiple times, apt-get binary does exist and runs without specifying the shell. Same error occurs when I run standard Linux commands e.g `curl` or `mv`  .
According to the docs, array format runs all of the given command without the shell (https://containers.dev/implementors/json_reference/#formatting-string-vs-array-properties).

The only other way is pass a single string to the field which runs all of the commands, separated by `&&` operator like: `apt-get update && apt-get upgrade`  or pass shell script and execute it (my current implementation).

Reply:
@camilamacedo86 IMO, passing all commands separated by `&&` in single string may resolve the issue but:

1. It will result in poor readability, and (potentially) bad debugging experience.
2. You can't break the string into multiple lines using `\` character.
3. `golangci-lint` will through an error related with single line character limit while linting. I saw this error when I was implementing the Dockerfile template for scaffolding.

Reply:
> golangci-lint will through an error related with single line character limit while linting. I saw this error when I was implementing the Dockerfile template for scaffolding.

We can ignore `golangci-lint ` checks for specific places
You can use `nolint: lll` you can check it in the code for other places and that is fine. 

> golangci-lint will through an error related with single line character limit while linting. I saw this error when I was implementing the Dockerfile template for scaffolding.

Do we really need to use apt get?
If so, and that does not work with Golang base image we can use the ubuntu base image.

Reply:
1. apt-get isn't necessary
2. devcontainer throws error for every command like mv, curl, chmod etc. Haven't been able to any root cause yet. I've also opened a new [issue](https://github.com/devcontainers/spec/issues/491) on the official repo just now.

Reply:
My bet is that it is permissions issues
The user used might need to be root.